contracts/contracts/strategies/aerodrome/AerodromeAMOStrategy.sol

```diff
@@ -0,0 +1,959 @@
+        if (_amountToSwap > 0) {
+            _swapToDesiredPosition(_amountToSwap, _minTokenReceived, _swapWeth);
+        }
+        // calling check liquidity early so we don't get unexpected errors when adding liquidity
+        // in the later stages of this function
+        _checkForExpectedPoolPrice();
+
+        _addLiquidity();
+        _checkForExpectedPoolPrice();
+    }
+
+    // slither-disable-end reentrancy-no-eth
+
+    /**
+     * @dev Decrease partial or all liquidity from the pool.
+     * @param _liquidityToDecrease The amount of liquidity to remove expressed in 18 decimal point
+     */
+    function _removeLiquidity(uint256 _liquidityToDecrease)
+        internal
+        gaugeUnstakeAndRestake
+    {
+        require(_liquidityToDecrease > 0, "Must remove some liquidity");
+
+        uint128 _liquidity = _getLiquidity();
+        // need to convert to uint256 since intermittent result is to big for uint128 to handle
+        uint128 liqudityToRemove = uint256(_liquidity)
+            .mulTruncate(_liquidityToDecrease)
+            .toUint128();
+
+        (uint256 _amountWeth, uint256 _amountOethb) = positionManager
+            .decreaseLiquidity(
+                // Both expected amounts can be 0 since we don't really care if any swaps
+                // happen just before the liquidity removal.
+                INonfungiblePositionManager.DecreaseLiquidityParams({
+                    tokenId: tokenId,
+                    liquidity: liqudityToRemove,
+                    amount0Min: 0,
+                    amount1Min: 0,
+                    deadline: block.timestamp
+                })
+            );
+
+        (
+            uint256 _amountWethCollected,
+            uint256 _amountOethbCollected
+        ) = positionManager.collect(
+                INonfungiblePositionManager.CollectParams({
+                    tokenId: tokenId,
+                    recipient: address(this),
+                    amount0Max: type(uint128).max, // defaults to all tokens owed
+                    amount1Max: type(uint128).max // defaults to all tokens owed
+                })
+            );
+
+        _updateUnderlyingAssets();
+
+        emit LiquidityRemoved(
+            _liquidityToDecrease,
+            _amountWeth, //removedWethAmount
+            _amountOethb, //removedOethbAmount
+            _amountWethCollected,
+            _amountOethbCollected,
+            underlyingAssets
+        );
+    }
+
+    /**
+     * @dev Perform a swap so that after the swap the ticker has the desired WETH to OETHb token share.
+     */
+    function _swapToDesiredPosition(
+        uint256 _amountToSwap,
+        uint256 _minTokenReceived,
+        bool _swapWeth
+    ) internal {
+        IERC20 _tokenToSwap = IERC20(_swapWeth ? WETH : OETHb);
+        uint256 _balance = _tokenToSwap.balanceOf(address(this));
+
+        if (_balance < _amountToSwap) {
+            // if swapping OETHb
+            if (!_swapWeth) {
```

Comment:
⚪️ We should use the if -> fail pattern here.

Reply:
nice catch: https://github.com/OriginProtocol/origin-dollar/pull/2184/commits/8fb4dd10a6cfc22a40bbc8a64ac27a9ab7908777